task/mgmt-gateway/src/main.rs

```diff
@@ -173,6 +178,22 @@ impl UsartHandler {
         sys_set_timer(None, TIMER_IRQ);
     }
 
+    fn drain_flushed_data(&mut self, n: usize) {
+        self.from_rx.drain(..n);
```

Comment:
Under the hood, this is going to `ptr::copy` the remaining data to shift it over by `n` ([source](https://docs.rs/arrayvec/latest/src/arrayvec/arrayvec.rs.html#983)), so it could do a bunch of extra work!

It seems like the "correct" data structure here is a circular buffer, although that makes `serialize_with_trailing_data` a little trickier, because the trailing data could be split over two slices.

Do you think it's worth fixing this now, or leaving a TODO?

Reply:
[`heapless::Deque`](https://docs.rs/heapless/latest/heapless/struct.Deque.html) looks promising (and already has an `as_slices` method!)

Reply:
Switched to `heapless::Deque` in 215a1a02a33f85858200b98a64e90fd99f56039d, replacing all the drains with loops over `pop_front()`s and rewriting the bit around `available_rx_space`. It's a big enough change to probably be worth a second look, if you don't mind.